aip/client-libraries/4236.md

```diff
@@ -0,0 +1,36 @@
+---
+id: 4236
+state: reviewing
+created: 2024-03-25
+---
+
+# Version-aware clients
+
+APIs may choose to annotate services with [google.api.api_version]. If 
+[google.api.api_version] is specified, version-aware clients **must** include
+the value of [google.api.api_version] in the request to the API. This allows
+services to abide by the schema and behavior of the service at the time that
+this API version was deployed. The format of the API version must be treated
+as opaque by clients. Services may use a format with an apparent structure,
+but clients must not rely on this to determine components within an API version,
+or attempt to construct other valid API versions.
+
+### Expected Generator and Client Library Behavior
+
+If a service is annotated with [google.api.api_version], client library
+generators **must** include the version in requests sent to the API via
+the HTTP Header `X-Goog-Api-Version`.
+
+For HTTP Requests, an HTTP Query Parameter `$apiVersion` can be used
+instead of the `X-Goog-Api-Version` HTTP header.
+
+Requests which contain the API version, must include either an HTTP query
+parameter `$apiVersion` or HTTP header `X-Goog-Api-Version`, but not both.
+
```

Comment:
Thought about this in Java implementation and feel we should guide consistent behavior across languages.

```suggestion
Generated client library should not let users override the `X-Goog-Api-Version` HTTP header. This should be regardless if the service is annotated with [google.api.api_version].
```
Open to suggestions on extact wording.

Reply:
I'd prefer not to add that, personally. There's only so much you can do to protect users who are determined to do weird things. The chances of someone *accidentally* setting that header are very slim - and if someone really wants to do so, they can always send a request directly from an HTTP client. We definitely shouldn't provide any way of specifically overriding the header - but I don't think "code that allows headers to be specified" should be modified to try to prevent this.

Reply:
> There's only so much you can do to protect users who are determined to do weird things.

I agree on this. On the other hand, it also seems valuable to give user usable feedback if we don't intend have this overriden. 

To add a bit context in Java, we already have some logic trying to resolve conflict between internal header set by defaults and user headers. I am inclined to add logic to throw exception when user attempt to add a header with key "x-goog-api-version" ([draft pr](https://github.com/googleapis/sdk-platform-java/pull/2745/files)). Alternative to this is no change to this existing logic, which will throw exception only when internal header has this key (that's when service has a api version). I don't feel too strongly, but the first one gives a bit more consistent feedback.

I realize this is only dealing with an edge case, but also feel that this case is language agnostic, so wanted to check if we want to deal with it consistently across languages.

WDTY?

Reply:
I think if Java wants to attempt to stop the user from doing it, that's fine - I don't think it should be in the AIP.